Feature: segmented profiles

murilommen opened the pull request.

murilommen committed:
updating whylogs dependency

murilommen committed:
add resultset flow for segments

murilommen committed:
fixes to make example actually run

murilommen committed:
whole pipeline runs now

murilommen committed:
almost working model

For some reason, Dataflow is not understanding the segmentation part OR
there is something going on with whylogs, as its being interpreted as
a DatasetProfileView instead of a SegmentedDatasetProfileView. Need to persist
the work so far, hence this commit

murilommen committed:
almost done with implementation, one piece missing

details:
* map reduce from read to profile is working
* to write to whylabs, we need the partition object as well
* the Partition object falls into "too complex object" case and can't
be type-inferred by Apache Beam
* had to solve something similar with the Segment object definition
* changed from dataclass to NamedTuple
* hint: -> defining our own Partition object, without the ColumnMapper might do the
trick

murilommen committed:
got segmentation working through the entire pipeline

murilommen committed:
cleaning code

murilommen committed:
update makefile

murilommen committed:
update poetry action version

murilommen committed:
modify actions to upload template

details: changing Makefile + publish.yaml to comprise for both templates
on the Github Actions runner and have it available on GCP

murilommen committed:
add missing make endpoint

murilommen committed:
fix make description

murilommen committed:
separating make targets on git actions

murilommen committed:
fix typos that led to wrong ref

murilommen committed:
missing backslash

murilommen committed:
fix job names on integ make target

murilommen committed:
minor fixes

murilommen committed:
fix column name for template examples

naddeoa commented on Makefile:
```diff
@@ -54,10 +61,10 @@ example_run_direct_table: requirements.txt ## Run the profile directly, job with
 		--input-mode=BIGQUERY_TABLE \
-		--input-bigquery-table=bigquery-public-data:hacker_news.comments \
-		--date-column=time_ts \
+		--input-bigquery-sql=bigquery-public-data.hacker_news.full \
```
This looks like a mistake, it should still be table right? -sql would be used for an actual sql query.

murilommen replied:
correct, fixed it now

naddeoa commented on Makefile:
```diff
@@ -54,10 +61,10 @@ example_run_direct_table: requirements.txt ## Run the profile directly, job with
+		--date-column=timestamp \
 		--date-grouping-frequency=Y \
-		--org-id=org-0 \
+		--org-id=org-fjx9Rz \
```
Are we changing out the org? We'll have to change out the api key too right?

murilommen replied:
I've changed it to test on my personal org (as i don't have access to org-0 in prod)

naddeoa replied:
ok, make sure to add me to it as well

naddeoa commented on Makefile:
```diff
@@ -67,17 +74,40 @@ example_run_direct_table: requirements.txt ## Run the profile directly, job with
 		--requirements_file=$(REQUIREMENTS)
 
+example_run_direct_segmented_table: JOB_NAME=$(NAME)
+example_run_direct_segmented_table: TEMPLATE=batch_segmented_bigquery
+example_run_direct_segmented_table: requirements.txt ## Run the profile directly, job without templatizing it first.
+	poetry run python src/ai/whylabs/templates/$(TEMPLATE).py \
+		--job_name="$(JOB_NAME)" \
+		--input-mode=BIGQUERY_TABLE \
+		--input-bigquery-table=bigquery-public-data.hacker_news.full \
+		--date-column=timestamp \
+		--date-grouping-frequency=Y \
+		--org-id=org-fjx9Rz \
+		--project=whylogs-359820 \
+		--region=$(REGION) \
+		--logging-level=DEBUG \
+		--output=gs://whylabs-dataflow-templates-tests/$(JOB_NAME)/profile \
+		--staging_location=gs://whylabs-dataflow-templates-tests/$(JOB_NAME)/staging \
+		--temp_location=gs://whylabs-dataflow-templates-tests/$(JOB_NAME)/tmp \
+		--tmp=gs://whylabs-dataflow-templates-tests/$(JOB_NAME)/profile \
+		--api-key=$(WHYLABS_API_KEY) \
+		--runner=DataflowRunner \
+		--dataset-id=model-12 \
+		--requirements_file=$(REQUIREMENTS) \
+		--segment_column=type
+
 example_run_direct_query: JOB_NAME=$(NAME)
 example_run_direct_query: TEMPLATE=batch_bigquery_template
 example_run_direct_query: requirements.txt ## Run the profile directly, job without templatizing it first.
 	poetry run python src/ai/whylabs/templates/$(TEMPLATE).py \
 		--job_name="$(JOB_NAME)" \
 		--input-mode=BIGQUERY_SQL \
 		--input-bigquery-sql='select * from `bigquery-public-data.hacker_news.comments`' \
-		--date-column=time_ts \
+		--date-column=timestamp \
```
Did this change? I don't see a change to the table name above

murilommen replied:
if we are using `bigquery-public-data.hacker_news.comments` then yes, it must be changed

naddeoa commented on README.md:
```diff
@@ -22,6 +22,7 @@ Here are the setup instructions under Ubuntu/WSL.
 - Set local python version to 3.8: `pyenv install 3.8 && pyenv local 3.8`
 - Explicitly set python version: `poetry env use $(which python)`
 - Install dependencies and generate a local virtual env for IDE/tooling: `poetry install`
+- For local development, make sure to authenticate to GCP with: `gcloud auth application-default login`
```
I think you actually have to authenticate with that in addition to `gcloud auth login`.

murilommen replied:
actually, only application-default logging in does it. it will prompt you for your password on the CLI though

murilommen committed:
add multiple columns for segmentation

murilommen committed:
merging both templates into 1

murilommen committed:
updating git actions file

murilommen committed:
format fix

murilommen committed:
updated variable naming

details:
- changed segment_column to segment_columns, as it might take more than
1 argument
- fixed an old reference to bigquery's public data

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -123,18 +235,54 @@ def process_batch(
+        yield batch
+
+
+# def serialize_segmented_profiles(input: Tuple[SegmentDefinition, DatasetProfileView]) -> List[bytes]:
+#     """
+#     This function converts a single ProfileIndex into a collection of
+#     serialized DatasetProfileViews so that they can subsequently be written
+#     individually to GCS, rather than as a giant collection that has to be
+#     parsed in a special way to get it back into a DatasetProfileView.
+#     """
+#     return [input[1].serialize()]
```
can remove

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -123,18 +235,54 @@ def process_batch(
 
 class ProfileIndexBatchConverter(ListBatchConverter):
-    def estimate_byte_size(self, batch: List[Tuple[str, DatasetProfileView]]) -> int:
+    def estimate_byte_size(self, batch: List[Tuple[Union[str, SegmentDefinition], DatasetProfileView]]) -> int:
```
I guess we can just make the key an Any since we apparently don't really care about its size. Either that or we should add the size to the size of the view. 

Kind of weird because its hard to know how much this even matters/gets used at all.

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -123,18 +235,54 @@ def process_batch(
+        writer = WhyLabsWriter(org_id=self.args.org_id, api_key=self.args.api_key, dataset_id=self.args.dataset_id)
+
+        for seg_def, view in batch:
+            self.logger.info("Writing segmented dataset profile to %s:%s.", self.args.org_id, self.args.dataset_id)
```
can we add the segment info here too

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -30,30 +43,53 @@ class TemplateArgs:
+    query: str
+
+
+class SegmentDefinition(NamedTuple):
+    segment_column: str
+    segment: Segment
+    date_group: str
```
Out of curiosity, do you know how this is used as a key? Does it stringify this via pickle or use the __str__?

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -98,6 +134,82 @@ def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[str, Data
+
+    @beam.DoFn.yields_elements
+    def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[SegmentDefinition, DatasetProfileView]]:
+        # If no segment column is informed, no point in running the processing
+        assert self.segment_columns is not None
```
Can this go in the init where its set

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -98,6 +134,82 @@ def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[str, Data
+        df[tmp_date_col] = pd.to_datetime(df[self.date_column])
+        grouped = df.set_index(tmp_date_col).groupby(pd.Grouper(freq=self.freq))
+
+        self.logger.info(f"Using {self.segment_columns} for segmentation")
+
+        seg_columns = self.segment_columns.split(",")
```
It would be good to move this validation stuff to outside of the actual pipeline so we can fail as early as possible before we start processing/reading any information. Can we pass the segmented columns into this class already parsed into a list of string, trimmed and everything?

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -98,6 +134,82 @@ def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[str, Data
+            # trim whitespaces on every string
+            segment_column = segment_column.strip()
+
+            column_segments = segment_on_column(segment_column)
```
Is this the right way of doing that? I thought you could just give a list of segments to whylogs and it would create the segments it needs to all in one go, and then you could get them all out at once. With this approach we end up processing the entire batch once for each segment defined.

naddeoa replied:
As a side note, we'll have to put a big warning in the docs about how this can blow up, even if we can figure out how to limit it to a single pass. Does whylogs throw or something if this is a continuous column that would explode the number of segment partitions?

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -30,30 +43,53 @@ class TemplateArgs:
 
-    def add_input(self, accumulator: DatasetProfileView, input: DatasetProfileView) -> DatasetProfileView:
-        return accumulator.merge(input)
+    def add_input(self, accumulator: DatasetProfileView, input: List[DatasetProfileView]) -> DatasetProfileView:
+        if len(input) == 0:
+            return accumulator
+
+        profile = DatasetProfile()
+        profile.track(pd.DataFrame.from_dict(input))
+        ret = accumulator.merge(profile.view())
+        return ret
 
     def merge_accumulators(self, accumulators: List[DatasetProfileView]) -> DatasetProfileView:
+        if len(accumulators) == 1:
+            # logger.info('Returning accumulator, only one to merge')
+            return accumulators[0]
+
         view: DatasetProfileView = DatasetProfile().view()
         for current_view in accumulators:
             view = view.merge(current_view)
```
looking at this again, I have no idea why its here. I think I might have just left this in from a previous iteration of this class? AFAIK, this is just used as a dataclass right? Can we just delete all of these methods?

naddeoa commented on metadata/batch_bigquery_template_metadata.json:
```diff
@@ -59,6 +59,12 @@
       "label": "Date column"
     },
+    {
+      "name": "segment_column",
```
this should be plural with a validation regex for a json list, or a fancier one that is a json list or a string if you want to allow both formats. I don't feel strongly there.

JSON list or csv, whatever is consistent. If there is already json in the template somewhere then I'd say keep it json.

murilommen committed:
fix broken test

murilommen committed:
Minor fixes:

- updated segment_columns on metadata.json
- using a regex as a validation pattern for the template
- cleaned up InputBigQueryTable dataclass
- remove unused commented code
- changed typing on ProfileBatchConverter

murilommen committed:
attempt to fix regex pattern

murilommen committed:
move segment_columns parsing to init

murilommen committed:
fix regex

murilommen committed:
making one why.log call for all segments

murilommen committed:
format fix

murilommen committed:
add logger to segments

murilommen committed:
naming fixes

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -30,30 +43,31 @@ class TemplateArgs:
+
+class SegmentDefinition(NamedTuple):
+    segment_column: str
+    segment: Segment
```
@jamie256 we're confused about what the tuple represents in this object. It seems to just be a single value with the way that we're using segmentation here.

naddeoa commented on src/ai/whylabs/templates/batch_bigquery_template.py:
```diff
@@ -98,6 +111,82 @@ def process_batch(self, batch: List[Dict[str, Any]]) -> Iterator[Tuple[str, Data
+            if len(dataframe) == 0:
+                continue
+
+            result_set = why.log(df, schema=DatasetSchema(segments=multi_column_segments))
```
small thing, we can just make the schema once

naddeoa commented on metadata/batch_bigquery_template_metadata.json:
```diff
@@ -59,6 +59,15 @@
+      "label": "Segment columns",
+      "isOptional": true,
+      "regexes": [
+        "[^,]+$"
```
I think this is what you want: `^[^,\n]*(,[^,\n]+)*$` (thx chatgpt)

The one that you have says "any sequence besides comma". But you really want "any sequence of any sequence besides comma, separated by commas".

murilommen committed:
addressing pr comments

murilommen committed:
small fixes

details:
- made a make target that will COALESCE null values with `False`
for multiple segmented column values
- changed SegmentDefinition.segment_column to segment_columns, as it might take more
than 1
